src/test-explorer/test-manager.ts

```diff
@@ -0,0 +1,163 @@
+    if (isDisabled) {
+      this.disable();
+    }
+    this.testController.resolveHandler = async (item?: vscode.TestItem) => {
+      if (item != null) {
+        return;
+      } else {
+        await this.discoverTestSuites();
+      }
+    };
+
+    this.testController.createRunProfile(
+      "Run",
+      TestRunProfileKind.Run,
+      (request, token) => runHandler(this.testController, true, request, token),
+      true
+    );
+
+    this.testController.createRunProfile(
+      "Debug",
+      TestRunProfileKind.Debug,
+      (request, token) =>
+        runHandler(this.testController, false, request, token),
+      false
+    );
+  }
+
+  async enable(): Promise<void> {
+    this.isDisabled = false;
+    await this.discoverTestSuites();
+  }
+
+  /**
+   * Disables test manager, also it deletes all discovered test items in order to hide gutter icons.
+   */
+  disable(): void {
+    this.testController.items.forEach((item) =>
+      this.testController.items.delete(item.id)
+    );
+    this.isDisabled = true;
+  }
+
+  discoverTestSuites(): Promise<void> {
+    if (this.isDisabled) {
+      return Promise.resolve();
+    }
+
+    return this.client
+      .sendRequest(ExecuteCommandRequest.type, {
+        command: "discover-test-suites",
+      })
+      .then(
+        (value: SuiteDiscovery[]) => {
+          for (const { targetName, targetUri, discovered } of value) {
+            const rootNode = this.testController.createTestItem(
+              targetName,
+              targetName
+            );
+            createTestItems(
+              this.testController,
+              discovered,
+              rootNode,
+              targetName,
+              targetUri
+            );
+            const data: TestItemMetadata = {
+              kind: "project",
+              targetName,
+              targetUri,
+            };
+            testCache.setMetadata(rootNode, data);
+            this.testController.items.add(rootNode);
+          }
+        },
+        (err) => console.error(err)
+      );
+  }
+}
+
+/**
+ * Create TestItems from the given @param discoveredArray and add them to the @param parent
+ */
+function createTestItems(
+  testController: vscode.TestController,
+  discoveredArray: TestDiscoveryResult[],
+  parent: vscode.TestItem,
+  targetName: TargetName,
+  targetUri: TargetUri
+) {
+  for (const discovered of discoveredArray) {
+    if (discovered.kind === "suite") {
+      const { className, location, fullyQualifiedName } = discovered;
+      const parsedUri = vscode.Uri.parse(location.uri);
+      const parsedRange = toVscodeRange(location.range);
+      const testItem = testController.createTestItem(
+        fullyQualifiedName,
+        className,
+        parsedUri
+      );
+      testItem.range = parsedRange;
+      const data: TestItemMetadata = {
+        kind: "suite",
+        targetName,
+        targetUri,
+      };
+      testCache.setMetadata(testItem, data);
```

Comment:
Why do we need the cache actually?

Reply:
`vscode.TestItem` doesn't contain any additional field for metadata. It's basically just an id, label, location and children. It's better to keep additional & necessary data elsewhere. Maybe the `cache` is not the best word for it, but I cannot think of something better.